src/Password/Argon2i.php

```diff
@@ -0,0 +1,197 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-crypt for the canonical source repository
+ * @copyright Copyright (c) 2005-2018 Zend Technologies USA Inc. (https://www.zend.com)
+ * @license   https://github.com/zendframework/zend-crypt/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\Crypt\Password;
+
+use Traversable;
+use Zend\Stdlib\ArrayUtils;
+
+use const PASSWORD_ARGON2I;
+use const PHP_VERSION_ID;
+
+use function is_array;
+use function password_hash;
+use function password_verify;
+use function strtolower;
+
+/**
+ * Argon2i algorithm using password_hash() function of PHP
+ */
+class Argon2i implements PasswordInterface
+{
+    /**
+     * Maximum memory (in kibibytes) that may be used to compute the Argon2 hash
+     *
+     * @var int|null
+     */
+    protected $memoryCost;
+
+    /**
+     * Maximum amount of time it may take to compute the Argon2 hash
+     *
+     * @var int|null
+     */
+    protected $timeCost;
+
+    /**
+     * Number of threads to use for computing the Argon2 hash
+     *
+     * @var int|null
+     */
+    protected $threads;
+
+    /**
+     * @var string
+     */
+    protected $salt;
+
+    /**
+     * Constructor
+     *
+     * @param array|Traversable $options
+     * @throws Exception\InvalidArgumentException
+     */
+    public function __construct($options = [])
+    {
+        if (PHP_VERSION_ID < 70200) {
+            throw new Exception\InvalidArgumentException(
+                'The Argon2i password hash is only nativly available on PHP7.2+'
+            );
+        }
+
+        if (! empty($options)) {
+            if ($options instanceof Traversable) {
+                $options = ArrayUtils::iteratorToArray($options);
+            }
+
+            if (! is_array($options)) {
+                throw new Exception\InvalidArgumentException(
+                    'The options parameter must be an array or a Traversable'
+                );
+            }
+
+            foreach ($options as $key => $value) {
+                switch (strtolower($key)) {
+                    case 'memory_cost':
+                        $this->setMemoryCost($value);
+                        break;
+                    case 'time_cost':
+                        $this->setTimeCost($value);
+                        break;
+                    case 'threads':
+                        $this->setThreads($value);
+                        break;
+                }
+            }
+        }
+    }
+
+    /**
+     * Returns the maximum memory (in kibibytes) that may be used to compute the Argon2 hash
+     *
+     * @return int
+     */
+    public function getMemoryCost()
+    {
+        return $this->memoryCost;
+    }
+
+    /**
+     * Sets the maximum memory (in kibibytes) that may be used to compute the Argon2 hash
+     *
+     * @param int $memoryCost
+     *
+     * @return Argon2i
+     */
+    public function setMemoryCost($memoryCost)
```

Comment:
@Ocramius I think we still need this in order to be compliant with the previous implementations like `Zend\Crypt\Password\Bcrypt`, `Zend\Crypt\Password\BcryptSha` and `Zend\Crypt\Password\Apache`.
Even, because we can add some check on setter to verify that the parameter is correct with a detailed error message if not. IMHO, this is actually one of the advantages to use a wrapper class like `Argon2i` instead of the PHP bult-in function.